src/reset_reason.rs

```diff
@@ -0,0 +1,59 @@
+use core::{
+    convert::{TryFrom, TryInto},
+    fmt,
+};
+use esp_idf_sys::*;
+
+#[derive(Debug)]
+pub enum ResetReason {
+    Software,
+    ExternalPin,
+    Watchdog,
+    Sdio,
+    Panic,
+    InterruptWatchdog,
+    PowerOn,
+    Unknown,
+    Brownout,
+    TaskWatchdog,
+    DeepSleep,
+}
+
+#[derive(Debug)]
+pub struct InvalidResetReason(esp_reset_reason_t);
+
+#[cfg(feature = "std")]
+impl std::error::Error for InvalidResetReason {}
+
+impl fmt::Display for InvalidResetReason {
```

Comment:
I think we should handle ALL reset reasons, and if there is an unknown reset reason, we should panic, rather than erroring out. How is it even possible to have `InvalidResetReason` besides a bug in the code below, which might be non-exhaustive? Bugs should be panics (or `unreachable!()`) IMO.

Reply:
This sounds reasonable to me. It does suggest that we may want to change `esp-idf-sys`'s bindgen to mark `esp_reset_reason_t` as a [rustified_enum](https://docs.rs/bindgen/latest/bindgen/struct.Builder.html#method.rustified_enum).

For now, I'll tweak this to `panic!()` on the unhandled part of the u32 range in `reset_reason()`. Unless folks are open to the esp-idf-sys change? Thoughts?